### PR TITLE
IPFilter caching and style changes

### PR DESCRIPTION
**Pull request by gkatsikas**
To see the actual changes introduced by caching, please check the first commit only.
The efficiency of the caching could be further improved if we use the RSS hash directly from the hardware, but this is left as future work :p

### CODE REVIEW

File: elements/ip/ipfilter.hh [hunk 1]
```diff
@@ -323,28 +372,55 @@ inline bool
...
+        if (caching) {
+            (*_cache_misses_nb.findp(name()))++;
+        }
+        // common case never checks packet length
+        return length_checked_match(zprog, p, packet_length);
+    }
+
+    // Caching enabled
+    if (caching) {
+        if (_last_flow_id[name()]) {
```
Comment (tbarbette) [inline elements/ip/ipfilter.hh]: Er... multiple hashtable search in the fast path? :p Why not a per-element cache?

Comment (gkatsikas) [reply elements/ip/ipfilter.hh]: Good catch, thanks! I was already fixing the problem as I realized myself how stupid it was :p

Comment (gkatsikas) [reply elements/ip/ipfilter.hh]: Do not merge yet, as I want to test the performance of caching with my new testie.
I will give you the green flag once I finish.
Of course, feel free to propose more improvements ;)

Comment (gkatsikas) [reply elements/ip/ipfilter.hh]: Performance looks good in general. With a small number of rules, there is a very light impact of caching on the latency, but with a large number of rules the performance gains can be substantial. Green flag from me :p

**Comment by tbarbette** [comment]
Yep seems good! Thanks !

**Pull request merged**